app/uk/gov/hmrc/gform/save4later/FormMongoCache.scala

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2021 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.gform.save4later
+
+import akka.http.scaladsl.model.StatusCodes
+import org.slf4j.{ Logger, LoggerFactory }
+import play.api.libs.json.{ Format, Json }
+import uk.gov.hmrc.crypto.CryptoWithKeysFromConfig
+import uk.gov.hmrc.gform.sharedmodel.form.{ Form, FormId, FormIdData }
+import uk.gov.hmrc.http.{ HeaderCarrier, UpstreamErrorResponse }
+import uk.gov.hmrc.mongo.cache.{ DataKey, MongoCacheRepository }
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+class FormMongoCache(mongoCacheRepository: MongoCacheRepository[String], jsonCrypto: CryptoWithKeysFromConfig)(implicit
+  ec: ExecutionContext
+) extends FormPersistenceAlgebra[Future] {
+
+  implicit val formatFormEncrypted: Format[Form] = EncyryptedFormat.formatEncrypted[Form](jsonCrypto)(Form.format)
+
+  private val formDataKey: DataKey[Form] = DataKey("form")
+
+  private val logger: Logger = LoggerFactory.getLogger(getClass)
+
+  override def find(formId: FormId)(implicit hc: HeaderCarrier): Future[Option[Form]] = mongoCacheRepository
+    .get[Form](formId.value)(formDataKey)
+
+  override def get(formId: FormId)(implicit hc: HeaderCarrier): Future[Form] = find(formId) map {
+    case None =>
+      throw UpstreamErrorResponse(
+        s"Not found 'form' for the given id: '${formId.value}'",
+        StatusCodes.NotFound.intValue
+      )
+    case Some(form) =>
+      logger.debug(Json.prettyPrint(Json.toJson(form)) + "GETFORM")
+      form
+  }
+
+  override def get(formIdData: FormIdData)(implicit hc: HeaderCarrier): Future[Form] = get(formIdData.toFormId)
+
+  override def upsert(formId: FormId, form: Form)(implicit hc: HeaderCarrier): Future[Unit] =
+    mongoCacheRepository
+      .put(formId.value)(formDataKey, form)
+      .map(_ => ())
```

Comment:
Is this safe, what if write failed?

Reply:
`put` returns a CacheItem, which contains details of the object insert in a success scenario, which we dont need. All failures will throw exception and cause Future to fail. We dont need anything from CacheItem

Reply:
Ok